### DISCUSSION TITLE
How to make heal skill on 5x5 range

### DISCUSSION THREAD

**Discussion by armoha** [armoha/euddraft#61] [Q&A] [answered]
Answer of https://cafe.naver.com/edac/111879 = https://cafe.naver.com/edac/111841

### Choice 1. Button
**Should we use primary order button? (Hold, Carrier Stop, Reaver Stop, ...)\
Or should we use secondary order button? (Train, Cloak, Rally, ...)**
- Major difference: Primary Order will cancel existing order. Secondary order does not interfere primary order.
- Some orders do not work on multiple selection (Train, Build, ...)

### Choice 2. Spatial data structure
I'm going to check simple rectangular position but it would be clever to use spatial data structure, like simple grid or quad tree, to avoid checking all units and instead check units only in boundary.

### Choice 3. How to check unit position
We'll use simple 4 DeathsX instead of Bring. CBringEPD is too slow to run on multiple units because it backups unitID, sets unitID to unique one, runs Bring and restores unitID, *on every unit to check*. If you use location, you must synchronize Location Coordinates and Position-checking conditions correctly and it would be nice if the synchronization only runs minimally on-demand.

### Choice 4. Prevent HP from exceeding maximum
SC gracefully handles this for shield point and units with hp regeneration. Turn on HP regeneration for simple reasoning. If you don't want HP regeneration, you need to build efficient data structure to load max hp by unit kind.

**Comment by armoha** [comment, marked as answer]
## epScript example)
- **Zergling** has heal skill, which costs **20 Energy** and heals **Hydralisks** by **10 HP** at 5x5 square, and uses **Train Broodling button** under the scenes.
- **Hydralisk** has HP regeneration so we'll ignore max hp.
```js
// BuildCheckConst : https://cafe.naver.com/edac/98602
import bcc.BuildCheckConst;

function healSkill() {
    var healCenter = 0;
    var healPlayer;

    // we'll cache unit selections because they're seldom changed 
    const ptrCache = EUDVArray(8 * 12)();
    const epdCache = EUDVArray(8 * 12)();
    const vr1, vr2 = EUDVArrayReader(), EUDVArrayReader();
    var ptr, epd;
    DoActions(
        vr1.seek(ptrCache, EPD(ptrCache), ptr), 
        vr2.seek(epdCache, EPD(epdCache), epd));

    const selectEPD = EPD(0x6284E8);
    var select = selectEPD;
    while (select < selectEPD + 8 * 12) {
        vr1.read();
        vr2.read();
        if (!MemoryEPD(select, Exactly, ptr)) {
            ptr, epd = cunitepdread_epd(select);
            const pos = (select - selectEPD) * 18;
            const updatePtr = EPD(ptrCache) + 78 + pos;
            const updateEPD = EPD(epdCache) + 78 + pos;
            dwwrite_epd(updatePtr, ptr);
            dwwrite_epd(updateEPD, epd);
        }
        if (ptr == 0) continue;
        const unitKind = epd + 25;
        if (MemoryEPD(unitKind, Exactly, $U("Zerg Zergling"))) {
            const buildQueue = epd + 0x98/4, epd + 0x9C/4, epd + 0xA0/4;
            if (MemoryEPD(buildQueue[0], Exactly, 0xE400E4)
            && MemoryEPD(buildQueue[1], Exactly, 0xE400E4)
            && MemoryXEPD(buildQueue[2], Exactly, 0xE4, 0xFFFF)
            ) { // Zergling isn't training any unit
            } else {
                // Zergling is training some unit
                const buildReset = function() {
                    dwwrite_epd(buildQueue[0], 0xE400E4);
                    dwwrite_epd(buildQueue[1], 0xE400E4);
                    wwrite_epd(buildQueue[2], 0, 0xE4);
                };

                if (BuildCheckConst(epd, "Zerg Broodling")) {
                    // We *must* read owner since user can select unit owned by another player
                    const owner = maskread_epd(epd + 19, 15);
                    const energy = epd + 0xA0/4;
                    if (MemoryXEPD(energy, AtMost, (20 << 16) - 1, 0xFFFF0000)) {
                        buildReset();
                        SetCurrentPlayer(owner);
                        DisplayText("Not enough energy");
                    } else if (healCenter == 0) {
                        buildReset();
                        SetMemoryXEPD(energy, Subtract, 20 << 16, 0xFFFF0000);
                        healPlayer = owner;
                        healCenter = epd + 0x28/4;
                    }
                }
            }
        }
        EUDSetContinuePoint();
        select += 1;
    }

    if (healCenter == 0) return;
    // Should we cache healCenter? I guess not...
    const healX = maskread_epd(healCenter, 0xFFFF);
    const healY = maskread_epd(healCenter, 0xFFFF0000);
    const healCond = DeathsX(CurrentPlayer, AtLeast, 0, 0, 0xFFFF),
        DeathsX(CurrentPlayer, AtMost, 0, 0, 0xFFFF),
        DeathsX(CurrentPlayer, AtLeast, 0, 0, 0xFFFF0000),
        DeathsX(CurrentPlayer, AtMost, 0, 0, 0xFFFF0000);
    dwwrite(healCond[0] + 8, healX);
    dwwrite(healCond[1] + 8, healX + 5 * 16);
    dwwrite(healCond[2] + 8, healY);
    dwwrite(healCond[3] + 8, healY + 5 * 16 * 65536);
    DoActions(  // prevent wrap-around at map boundary
        SetMemory(healCond[0] + 8, Subtract, 5 * 16),
        SetMemory(healCond[2] + 8, Subtract, 5 * 16 * 65536)
    );

    foreach(ptr, epd : EUDLoopUnit2()) {
        const unitKind = epd + 25;
        if (MemoryEPD(unitKind, Exactly, $U("Zerg Hydralisk"))) {
            SetCurrentPlayer(epd + 0x28/4);
            if (MemoryXEPD(epd + 19, Exactly, healPlayer, 0xFF)
            && healCond) {
                dwadd_epd(epd + 2, 10 * 256);
            }
        }
    }
}
```

**Reply by TAEKHi** [reply]
Hello!
I modified the ButtonSet and applied the code as intended in the text, but it does not apply to "Zergling". Please let me know what I need to do more.
(FYI, I think ":" is an error in the code. I've solved that.)
Thank you for teaching me :)

**Reply by armoha** [reply]
@TAEKHi You need to set broodling train requirement, and set broodling's cost and supply to 0.

**Reply by TAEKHi** [reply]
![1](https://user-images.githubusercontent.com/104509268/165534889-c6a3369b-5dd8-4442-90b7-e3c403cd0eab.jpg)
![2](https://user-images.githubusercontent.com/104509268/165534897-b3482a61-6bf6-412b-8386-e8526b8750dd.jpg)
![3](https://user-images.githubusercontent.com/104509268/165534903-56373d4f-9d2b-4e48-9b5d-7aee47801af1.jpg)
![4](https://user-images.githubusercontent.com/104509268/165534906-cd45c220-98da-409f-9f8d-cb952dee1fbd.jpg)

T^T
I'll give you a file if you want.

**Reply by armoha** [reply]
@TAEKHi You need to call `healSkill();` function repeatedly at either `beforeTriggerExec` or `afterTriggerExec`

**Reply by TAEKHi** [reply]
Thank you sooo much😘

**Reply by TAEKHi** [reply]
@armoha 
sir  I'm sorry to bother you.
LASTLY, I have a question.
Could you make sure that always heals without Training the "broodling"?

I don't know what to take out of the code above..

**Reply by armoha** [reply]
You mean passive healing aura ability?
See https://github.com/armoha/euddraft/discussions/62

**Reply by TAEKHi** [reply]
Yes! you understand very well.
